File: design-pattern-java/app/src/test/java/design/pattern/java/patterns/structural/proxy/ProxyTest.java
package design.pattern.java.patterns.structural.proxy;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ProxyTest {
  class ServiceMock implements Service {
    private Service service;
    private Boolean wasCalled;
    private Integer times;

    ServiceMock(Service service) {
      super();
      this.service = service;
      this.wasCalled = false;
      this.times = 0;
    }

    @Override
    public String addPrefix(String word) {
      this.wasCalled = true;
      this.times++;
      return this.service.addPrefix(word);
    }

    public Boolean hasBeenCalled() {
      return this.wasCalled;
    }

    public Boolean hasBeenCalledTimes(Integer num) {
      return this.times == num;
    }
  }

  @Test
  public void testShouldCallServiceWhenSameMethodProxyIsCalled() {
    var mock = new ServiceMock(new AddCustomPrefix(""));
    var proxy = new ServiceProxy(mock);
    proxy.addPrefix("word");
    Assertions.assertTrue(mock.hasBeenCalled());
  }

  @Test
  public void testShouldCallService1WhenMthIsCalledMultTimes() {
    var mock = new ServiceMock(new AddCustomPrefix("-"));
    var proxy = new ServiceProxy(mock);
    proxy.addPrefix("1");
    proxy.addPrefix("1");
    proxy.addPrefix("1");
    Assertions.assertTrue(mock.hasBeenCalledTimes(1));
  }

  @Test
  public void testShouldReturnCachedValueWhenSameParamIsSent() {
    var service = new AddCustomPrefix("");
    var proxy = new ServiceProxy(service);
    var first = proxy.addPrefix("word");
    var second = proxy.addPrefix("word");
    var third = proxy.addPrefix("word");
    Assertions.assertTrue(first == second && second == third);
  }

}

File: design-pattern-jvm/app/src/test/kotlin/design/pattern/kotlin/patterns/structural/proxy/ProxyTest.kt
package design.pattern.kotlin.patterns.structural.proxy

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ProxyTestKt {
    internal inner class ServiceMock(private val service: Service) : Service {
        private var wasCalled = false
        private var times = 0
        override fun addPrefix(word: String): String {
            wasCalled = true
            times++
            return service.addPrefix(word)
        }

        fun hasBeenCalled(): Boolean {
            return wasCalled
        }

        fun hasBeenCalledTimes(num: Int): Boolean {
            return times == num
        }
    }

    @Test
    fun testShouldCallServiceWhenSameMethodProxyIsCalled() {
        val mock: ServiceMock = ServiceMock(AddCustomPrefix(""))
        val proxy = ServiceProxy(mock)
        proxy.addPrefix("word")
        Assertions.assertTrue(mock.hasBeenCalled())
    }

    @Test
    fun testShouldCallService1WhenMthIsCalledMultTimes() {
        val mock: ServiceMock = ServiceMock(AddCustomPrefix("-"))
        val proxy = ServiceProxy(mock)
        proxy.addPrefix("1")
        proxy.addPrefix("1")
        proxy.addPrefix("1")
        Assertions.assertTrue(mock.hasBeenCalledTimes(1))
    }

    @Test
    fun testShouldReturnCachedValueWhenSameParamIsSent() {
        val service = AddCustomPrefix("")
        val proxy = ServiceProxy(service)
        val first = proxy.addPrefix("word")
        val second = proxy.addPrefix("word")
        val third = proxy.addPrefix("word")
        Assertions.assertTrue(first === second && second === third)
    }
}
